pandas/core/groupby/generic.py

```diff
@@ -1569,6 +1571,161 @@ def func(df):
+            )
+
+        with self._group_selection_context():
+            df = self.obj
+
+            in_axis_names = {
+                grouping.name for grouping in self.grouper.groupings if grouping.in_axis
+            }
+            if isinstance(self._selected_obj, Series):
+                name = self._selected_obj.name
+                keys = [] if name in in_axis_names else [self._selected_obj]
+            else:
+                keys = [
+                    # TODO: Is there a better way to get the Series for the grouper?
+                    self._selected_obj.iloc[:, idx]
+                    for idx, name in enumerate(self._selected_obj.columns)
+                    if name not in in_axis_names
+                ]
+
+            if subset is not None:
+                clashing = set(subset) & set(in_axis_names)
+                if clashing:
+                    raise ValueError(
+                        f"Keys {clashing} in subset cannot be in "
+                        "the groupby column keys"
+                    )
+
+            groupings = list(self.grouper.groupings)
+            for key in keys:
+                grouper, _, _ = get_grouper(
+                    df,
+                    key=key,
+                    axis=self.axis,
+                    level=self.level,
+                    sort=self.sort,
+                    mutated=self.mutated,
+                    dropna=dropna,
+                )
+                groupings += list(grouper.groupings)
+
+            # Take the size of the overall columns
+            gb = df.groupby(
+                groupings,
+                sort=self.sort,
+                observed=self.observed,
+                dropna=self.dropna,
+            )
+            result = cast(Series, gb.size())
+
+            if normalize:
+                # Normalize the results by dividing by the original group sizes.
+                # We are guaranteed to have the first N levels be the
+                # user-requested grouping.
+                # TODO: Is there a better way to subset levels from a MultiIndex?
+                levels = list(range(len(self.grouper.groupings), result.index.nlevels))
+                indexed_group_size = result.groupby(
+                    result.index.droplevel(levels),
+                    sort=self.sort,
+                    observed=self.observed,
+                    dropna=self.dropna,
+                ).transform("sum")
+
+                result /= indexed_group_size
+
+            if sort:
+                # Sort the values and then resort by the main grouping
+                index_level = range(len(self.grouper.groupings))
+                result = result.sort_values(ascending=ascending).sort_index(
+                    level=index_level, sort_remaining=False
+                )
+
+            if not self.as_index:
+                # Convert to frame
+                result = result.reset_index(name="proportion" if normalize else "count")
```

Comment:
Because of the possibility of duplicate columns, I'd guess you'd be better off running with as_index=False and then using `self._insert_inaxis_grouper_inplace(result)`

Reply:
Ah - this is wrong for multiple reasons (the method I mentioned does the opposite - adds the groupers to the columns. I think this is okay as-is. Can you do a check (upfront, before any computation) raise a NotImplementedError saying that as_index=False with duplicate columns is not supported.

Reply:
Running with as_index =False caused all the complexity with sort and normalize that you suggested that I remove! In any case, with my change to frame.py, reset_index() now works fine with duplicates.